### PR TITLE
[C++ Worker]Add gcs global state accessor

### PR DESCRIPTION
**Pull request by qicosmos**
## Why are these changes needed?

Add api: GetGlobalStateAccessor, 
We need this api to get placment group by name/id and get all placement groups like java:
https://github.com/ray-project/ray/blob/master/java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java#L209
https://github.com/ray-project/ray/blob/master/java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java#L214
https://github.com/ray-project/ray/blob/master/java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java#L219

Add api: WasCurrentActorRestarted, check the current actor was restarted or not.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests

### CODE REVIEW

**Comment by SongGuyang** [comment]
Can you add some comments why we need this API?

**Comment by qicosmos** [comment]
> Can you add some comments why we need this API?
Ok, i updated the pr description.

File: cpp/src/ray/runtime/abstract_ray_runtime.cc [hunk 1]
```diff
@@ -45,13 +44,17 @@ using ray::core::CoreWorkerProcess;
...
 
 std::shared_ptr<AbstractRayRuntime> AbstractRayRuntime::DoInit() {
   std::shared_ptr<AbstractRayRuntime> runtime;
   if (ConfigInternal::Instance().run_mode == RunMode::SINGLE_PROCESS) {
     runtime = std::shared_ptr<AbstractRayRuntime>(new LocalModeRayRuntime());
   } else {
-    ProcessHelper::GetInstance().RayStart(TaskExecutor::ExecuteTask);
+    global_state_accessor_ = nullptr;
+    ProcessHelper::GetInstance().RayStart(TaskExecutor::ExecuteTask,
+                                          &global_state_accessor_);
```
Comment (raulchen) [inline cpp/src/ray/runtime/abstract_ray_runtime.cc:57]: passing `gcs_global_state_accessor` to `RayStart` is weird. 
You can get the global state accessor from the CoreWorker instead.

Comment (qicosmos) [reply cpp/src/ray/runtime/abstract_ray_runtime.cc:57]: The global state accessor is not in CoreWorker, can't get it from coreworker.

File: cpp/src/ray/runtime/abstract_ray_runtime.cc [hunk 2]
```diff
@@ -242,5 +246,10 @@ void AbstractRayRuntime::ExitActor() {
   throw RayIntentionalSystemExitException("SystemExit");
 }
 
+const std::unique_ptr<ray::gcs::GlobalStateAccessor>
+    &AbstractRayRuntime::GetGlobalStateAccessor() {
```
[the remark this answers is not in the document]
Comment (qicosmos) [reply cpp/src/ray/runtime/abstract_ray_runtime.cc:250]: It used to get placement group information.
https://github.com/ray-project/ray/blob/master/java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java#L209

[the remark this answers is not in the document]
Comment (qicosmos) [reply cpp/src/ray/runtime/abstract_ray_runtime.cc:250]: It is a part of placement group.

File: cpp/src/ray/util/process_helper.cc [hunk 3]
```diff
@@ -106,6 +105,15 @@ void ProcessHelper::StopRayNode() {
   return;
 }
 
+std::unique_ptr<ray::gcs::GlobalStateAccessor> ProcessHelper::CreateGlobalStateAccessor(
+    const std::string &redis_address, const std::string &redis_password) {
+  std::unique_ptr<ray::gcs::GlobalStateAccessor> global_state_accessor = nullptr;
+  global_state_accessor.reset(
+      new ray::gcs::GlobalStateAccessor(redis_address, redis_password));
```
Comment (SongGuyang) [inline cpp/src/ray/util/process_helper.cc:112]: Use `std::make_unique` instead?

File: cpp/src/ray/runtime/abstract_ray_runtime.cc [hunk 4]
```diff
@@ -53,6 +54,17 @@ std::shared_ptr<AbstractRayRuntime> AbstractRayRuntime::DoInit() {
   } else {
     ProcessHelper::GetInstance().RayStart(TaskExecutor::ExecuteTask);
     runtime = std::shared_ptr<AbstractRayRuntime>(new NativeRayRuntime());
+    if (ConfigInternal::Instance().worker_type == WorkerType::DRIVER) {
```
Comment (SongGuyang) [inline cpp/src/ray/runtime/abstract_ray_runtime.cc:57]: We also need global_state_accessor in worker?

File: cpp/src/ray/runtime/abstract_ray_runtime.h [hunk 5]
```diff
@@ -103,6 +106,7 @@ class AbstractRayRuntime : public RayRuntime {
 
  private:
   static std::shared_ptr<AbstractRayRuntime> abstract_ray_runtime_;
+  static std::unique_ptr<ray::gcs::GlobalStateAccessor> global_state_accessor_;
```
Comment (SongGuyang) [inline cpp/src/ray/runtime/abstract_ray_runtime.h:109]: Can we make it non-static member of AbstractRayRuntime?

File: cpp/src/ray/util/process_helper.h [hunk 6]
```diff
@@ -37,11 +38,16 @@ class ProcessHelper {
...
+      const std::string &redis_address, const std::string &redis_password);
+
   ProcessHelper(ProcessHelper const &) = delete;
   void operator=(ProcessHelper const &) = delete;
 
  private:
   ProcessHelper(){};
 };
+
+std::string GetNodeIpAddress(const std::string &address = "8.8.8.8:53");
```
Comment (SongGuyang) [inline cpp/src/ray/util/process_helper.h:51]: Can we build a new file named util.cc for `GetNodeIpAddress `? It's a little weird to expose this API in process helper.

File: cpp/src/ray/util/util.cc [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    boost::asio::ip::address addr = socket.local_endpoint().address();
+    return addr.to_string();
+  } catch (std::exception &e) {
+    RAY_LOG(FATAL) << "Could not get the node IP address with socket. Exception: "
+                   << e.what();
+    return "";
+  }
+}
+}  // namespace internal
+}  // namespace ray
```
Comment (SongGuyang) [inline cpp/src/ray/util/util.cc:45]: Please Add a empty line at the end of file.

File: cpp/src/ray/util/util.h [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+///
+/// TODO(kfstorm): Make this function shared code and migrate Python & Java to use this
+/// function.
+///
+/// \param address The IP address and port of any known live service on the network
+/// you care about.
+/// \return The IP address by which the local node can be reached from the address.
+std::string GetNodeIpAddress(const std::string &address = "8.8.8.8:53");
+}  // namespace internal
+}  // namespace ray
```
Comment (SongGuyang) [inline cpp/src/ray/util/util.h:35]: Please Add a empty line at the end of file.

File: cpp/src/ray/util/process_helper.cc [hunk 9]
```diff
@@ -155,8 +128,8 @@ void ProcessHelper::RayStart(CoreWorkerOptions::TaskExecutionCallback callback)
     std::string session_dir = ConfigInternal::Instance().session_dir;
     if (session_dir.empty()) {
       if (!global_state_accessor) {
-        global_state_accessor.reset(new ray::gcs::GlobalStateAccessor(
-            redis_address, ConfigInternal::Instance().redis_password));
+        global_state_accessor = std::make_unique<ray::gcs::GlobalStateAccessor>(
```
Comment (SongGuyang) [inline cpp/src/ray/util/process_helper.cc:131]: Why don't use CreateGlobalStateAccessor here?

File: cpp/src/ray/runtime/abstract_ray_runtime.cc [hunk 10]
```diff
@@ -53,6 +52,16 @@ std::shared_ptr<AbstractRayRuntime> AbstractRayRuntime::DoInit() {
...
     ProcessHelper::GetInstance().RayStart(TaskExecutor::ExecuteTask);
     runtime = std::shared_ptr<AbstractRayRuntime>(new NativeRayRuntime());
+    auto redis_ip = ConfigInternal::Instance().redis_ip;
+    if (redis_ip.empty()) {
+      redis_ip = GetNodeIpAddress();
+    }
+    std::string redis_address =
+        redis_ip + ":" + std::to_string(ConfigInternal::Instance().redis_port);
+    runtime->global_state_accessor_ =
+        ProcessHelper::GetInstance().CreateGlobalStateAccessor(
```
Comment (SongGuyang) [inline cpp/src/ray/runtime/abstract_ray_runtime.cc:62]: Can we move this logic into NativeRayRuntime? "CreateGlobalStateAccessor" only works in native mode.

**Review by SongGuyang — APPROVED** [review]
LGTM.

**Pull request merged**